### PR TITLE
feat(core): introduce lua-resty-simdjson

### PR DESCRIPTION
**Pull request by chronolaw**
### Summary

lua-resty-simdjson has released the first version [`1.0.0`](https://github.com/Kong/lua-resty-simdjson/releases/tag/1.0.0).

KAG-3647

Summary:
* a new conf option `cluster_cjson` to switch between cjson and simdjson
* unit test for cjson compatibility

simdjson requires gcc 7.4+, it may not be compiled successfully on some Linux releases (aws-liunx-2). 

### Checklist

- [x] The Pull Request has tests
- [x] A changelog file has been created under `changelog/unreleased/kong` or `skip-changelog` label added on PR if changelog is unnecessary. [README.md](https://github.com/Kong/gateway-changelog/blob/main/README.md)
- [ ] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - PUT DOCS PR HERE

### Issue reference

Fix #_[issue number]_

### CODE REVIEW

**Comment by Water-Melon** [comment]
We need to fix the following issue: https://github.com/Kong/kong/actions/runs/10071146552/job/27840916315?pr=13421.

Add the new content indicated by the diff to the specified position in `fixtures/ubuntu-22.04-amd64.txt`.

**Comment by chronolaw** [comment]
> We need to fix the following issue: https://github.com/Kong/kong/actions/runs/10071146552/job/27840916315?pr=13421.
> 
> Add the new content indicated by the diff to the specified position in `fixtures/ubuntu-22.04-amd64.txt`.

Should we update other files like `ubuntu-20.04-amd64.txt` `debian-12-amd64.txt`?

**Comment by Water-Melon** [comment]
> > We need to fix the following issue: https://github.com/Kong/kong/actions/runs/10071146552/job/27840916315?pr=13421.
> > Add the new content indicated by the diff to the specified position in `fixtures/ubuntu-22.04-amd64.txt`.
> 
> Should we update other files like `ubuntu-20.04-amd64.txt` `debian-12-amd64.txt`?


Yes, we need to run the full matrix to verify all manifests and then update the files indicated in the failed jobs.

File: spec/01-unit/31-simdjson/01-cjson_compatibility_spec.lua [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+[[
+{
+    "glossary": {
+        "title": "example glossary",
+                "GlossDiv": {
+            "title": "S",
+                        "GlossList": {
+                "GlossEntry": {
+                    "ID": "SGML",
+                                        "SortAs": "SGML",
```
Comment (Water-Melon) [inline spec/01-unit/31-simdjson/01-cjson_compatibility_spec.lua:21]: the format alignment

File: spec/01-unit/31-simdjson/01-cjson_compatibility_spec.lua [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+      {"value": "Open", "onclick": "OpenDoc()"},
+      {"value": "Close", "onclick": "CloseDoc()"}
+    ]
+  }
+}}
+]],
+
+[[
+{"widget": {
+    "debug": "on",
```
Comment (Water-Melon) [inline spec/01-unit/31-simdjson/01-cjson_compatibility_spec.lua:53]: Some indentations use two spaces, while others use four spaces

File: spec/01-unit/31-simdjson/01-cjson_compatibility_spec.lua (hunk 1 above)
Comment (chronolaw) [reply spec/01-unit/31-simdjson/01-cjson_compatibility_spec.lua:21]: No, we will keep the original data format.

File: spec/01-unit/31-simdjson/01-cjson_compatibility_spec.lua (hunk 2 above)
Comment (chronolaw) [reply spec/01-unit/31-simdjson/01-cjson_compatibility_spec.lua:53]: No, we will keep the original data format.

**Review by Water-Melon — APPROVED** [review]

File: changelog/unreleased/kong/resty-simdjson.yml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+message: |
+  Introduced a new yieldable JSON library lua-resty-simdjson,
+  which will improve the latency dramatically.
```
Comment (outsinre) [inline changelog/unreleased/kong/resty-simdjson.yml:3]:
```suggestion
  Introduced a yieldable JSON library `lua-resty-simdjson` to replace the `cjson` library,
  which would improve the latency significantly.
```

File: kong/conf_loader/constants.lua [hunk 4]
```diff
@@ -499,6 +499,7 @@ local CONF_PARSERS = {
   cluster_use_proxy = { typ = "boolean" },
   cluster_dp_labels = { typ = "array" },
   cluster_rpc = { typ = "boolean" },
+  cluster_cjson = { typ = "boolean" },
```
Comment (outsinre) [inline kong/conf_loader/constants.lua:502]: How is this option used?

File: spec/01-unit/31-simdjson/01-cjson_compatibility_spec.lua [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+      assert.same(deep_sort(obj1), deep_sort(obj2))
+    end
+
+    parser:destroy()
+  end)
+
+  it("runs with separated interface", function ()
+
+    local dec = require("resty.simdjson.decoder").new()
+    local enc = require("resty.simdjson.encoder").new()
```
Comment (outsinre) [inline spec/01-unit/31-simdjson/01-cjson_compatibility_spec.lua:271]: What if we add a test on the yieldable `new(true)`?

Comment (chronolaw) [reply spec/01-unit/31-simdjson/01-cjson_compatibility_spec.lua:271]: See: https://github.com/Kong/lua-resty-simdjson/blob/main/t/04-yield.t

File: kong/conf_loader/constants.lua (hunk 4 above)
Comment (chronolaw) [reply kong/conf_loader/constants.lua:502]: It is a "hiden" option, like "legacy_events" "legacy_dns_client"

Comment (outsinre) [reply kong/conf_loader/constants.lua:502]: Seems I cannot find assignment to this var in this PR.

File: changelog/unreleased/kong/resty-simdjson.yml (hunk 3 above)
Comment (chronolaw) [reply changelog/unreleased/kong/resty-simdjson.yml]: Partly accepted, but in this PR we will not replace cjson completely.

File: spec/01-unit/31-simdjson/01-cjson_compatibility_spec.lua (hunk 5 above)
Comment (outsinre) [reply spec/01-unit/31-simdjson/01-cjson_compatibility_spec.lua:271]: How about, at Kong side, add a `new(true)` test.

File: kong/conf_loader/constants.lua (hunk 4 above)
Comment (chronolaw) [reply kong/conf_loader/constants.lua:502]: It is a env var.

File: kong/clustering/utils.lua [hunk 6]
```diff
@@ -164,4 +165,33 @@ function _M.is_dp_worker_process()
   return worker_id() == 0
 end
 
+
+-- encode/decode json with cjson or simdjson
+local ok, simdjson_dec = pcall(require, "resty.simdjson.decoder")
+if not ok or kong.configuration.cluster_cjson then
```
Comment (outsinre) [inline kong/clustering/utils.lua:171]: Should we give `cluster_cjson` a default value?

If `not ok`, and its default value is `nil`, this code forces `resty.simdjson`.

Comment (chronolaw) [reply kong/clustering/utils.lua:171]: added, thanks.

**Review by ADD-SP — APPROVED** [review]

File: kong/templates/kong_defaults.lua [hunk 7]
```diff
@@ -42,6 +42,7 @@ cluster_max_payload = 16777216
 cluster_use_proxy = off
 cluster_dp_labels = NONE
 cluster_rpc = off
+cluster_cjson = off
```
Comment (outsinre) [inline kong/templates/kong_defaults.lua:45]: Should be `true`?

**Review by outsinre — APPROVED** [review]

File: spec/01-unit/31-simdjson/01-cjson_compatibility_spec.lua (hunk 5 above)
Comment (chronolaw) [reply spec/01-unit/31-simdjson/01-cjson_compatibility_spec.lua:271]: It is possible but not urgent, we can add tests later.

File: kong/templates/kong_defaults.lua (hunk 7 above)
Comment (chronolaw) [reply kong/templates/kong_defaults.lua:45]: The default value `off` of `cluster_cjson` means than we will enable simdjson in clustering by default.

**Pull request merged**